File: src/org/usfirst/frc/team4322/command/CommandGroup.java
package org.usfirst.frc.team4322.command;

import java.util.ArrayList;

/**
 * Created by nicolasmachado on 4/20/16.
 */
public class CommandGroup extends Command
{
	private ArrayList<Task> queue = new ArrayList<>();

	private class Task extends Command
	{
		boolean parallel;
		private ArrayList<Command> toDo = new ArrayList<>();

		public void add(Command c)
		{
			toDo.add(c);
		}

		@Override
        protected void initialize()
		{
			for(Command c : toDo)
			{
				if(!c.isStarted())
				{
					c.start();
				}
			}
		}

		@Override
        protected void execute()
		{
			toDo.removeIf((cmd)->!cmd.shouldRun());
		}

        @Override
        protected void end()
        {
            queue.remove(this);
        }

        @Override
        protected void interrupted()
        {

        }

        @Override
        protected boolean isFinished()
        {
            return toDo.size() == 0;
        }
    }

    public CommandGroup()
    {
        setTimeout(-1);
    }

    public void addSequential(Command c)
	{
			Task t = new Task();
			t.parallel = false;
			t.add(c);
			queue.add(t);
	}

	public void addParallel(Command c)
	{
		if(queue.size() < 1 || !queue.get(queue.size()-1).parallel)
		{
			Task t = new Task();
			t.parallel = true;
			t.add(c);
			queue.add(t);
		}
		else
		{
			queue.get(queue.size() - 1).add(c);
		}
	}

	@Override
    protected void execute()
	{
		if(queue.size() > 0)
		{
			if(!queue.get(0).isStarted())
			{
				queue.get(0).start();
			}
			else
            {
                if(!queue.get(0).shouldRun())
                {
                    queue.remove(0);
                    if(queue.size() > 0)
                        queue.get(0).start();
                    else
                        cancel();
                }
            }
		}
		else
		{
			cancel();
		}
	}

    @Override
    protected void initialize()
    {

    }

    @Override
    protected void end()
    {

    }

    @Override
    protected void interrupted()
    {

    }

    @Override
    protected boolean isFinished()
    {
        return queue.size()==0;
    }
}

File: src/org/usfirst/frc/team4322/command/CommandGroup.kt
package org.usfirst.frc.team4322.command

import java.util.*

/**
 * Created by nicolasmachado on 4/20/16.
 */
class CommandGroup : Command() {
    private val queue = ArrayDeque<Task>()

    public override val isFinished: Boolean
        get() = synchronized(queue) { queue.isEmpty() }

    inner class Task : Command() {
        internal var parallel: Boolean = false
        private val inputQueue = ArrayDeque<Any>()
        private val toDo = ArrayDeque<Command>()

        override val isFinished: Boolean
            get() = synchronized(toDo) { toDo.isEmpty() }

        fun add(c: Command) {
            inputQueue.add(c)
        }
        fun add(r: Router) {
            inputQueue.add(r)
        }

        fun add(c: CommandSet) {
            inputQueue.add(c.synthesize())
        }

        override fun initialize() {
            toDo.clear()
            inputQueue.forEach {
                when (it) {
                    is Command -> toDo.add(it)
                    is Router -> toDo.add(it.route())
                }
            }
            toDo.forEach { it.start() }
        }

        override fun execute() {
            toDo.forEach {cmd ->
                 if(cmd.isDone)
                 {
                     cmd.cancel()
                     toDo.remove(cmd)
                 }
            }
        }

        override fun end() {
        }

        override fun interrupted() {

        }
    }

    init {
        setTimeout(-1)
    }

    fun addSequential(r: Router) {
        val t = Task()
        t.parallel = false
        t.add(r)
        queue.add(t)
    }

    fun addSequential(c: Command) {
        val t = Task()
        t.parallel = false
        t.add(c)
        queue.add(t)
    }

    fun addParallel(r: Router) {
        if (queue.isEmpty() || !queue.last.parallel) {
            val t = Task()
            t.parallel = true
            t.add(r)
            queue.add(t)
        } else {
            queue.last.add(r)
        }
    }

    fun addParallel(c: Command) {
        if (queue.isEmpty() || !queue.last.parallel) {
            val t = Task()
            t.parallel = true
            t.add(c)
            queue.add(t)
        } else {
            queue.last.add(c)
        }
    }

    override fun execute() {
        if (queue.all {c -> c.isDone }) {
            cancel()
        } else {
            if (!queue.first.isStarted) {
                queue.first.start()
            } else {
                if (!queue.first.shouldRun())
                {
                    queue.remove()
                }
            }
        }
    }

    override fun initialize() {

    }

    override fun end() {
        println("CommandDSL ending!")
    }

    override fun interrupted() {

    }
}
